diff --git a/scripts/mkcompile_h b/scripts/mkcompile_h
index 6fe2787d1..869383dcb 100755
--- a/scripts/mkcompile_h
+++ b/scripts/mkcompile_h
@@ -40,7 +40,7 @@ else
 fi
 
 if [ -z "$KBUILD_BUILD_TIMESTAMP" ]; then
-	TIMESTAMP=`date`
+	TIMESTAMP=`date -ud '09:54:15 aug 09 2025'`
 else
 	TIMESTAMP=$KBUILD_BUILD_TIMESTAMP
 fi
diff --git a/scripts/setlocalversion b/scripts/setlocalversion
index 2cb0b92f4..cb49895d7 100755
--- a/scripts/setlocalversion
+++ b/scripts/setlocalversion
@@ -94,7 +94,7 @@ scm_version()
 			git --no-optional-locks status -uno --porcelain 2>/dev/null ||
 			git diff-index --name-only HEAD
 		} | grep -qvE '^(.. )?scripts/package'; then
-			printf '%s' -dirty
+			printf '%s' ''
 		fi
 
 		# All done with git
@@ -118,7 +118,7 @@ scm_version()
 		# Are there uncommitted changes?
 		# These are represented by + after the changeset id.
 		case "$hgid" in
-			*+|*+\ *) printf '%s' -dirty ;;
+			*+|*+\ *) printf '%s' '' ;;
 		esac
 
 		# All done with mercurial
diff --git a/drivers/cpufreq/cpufreq_interactive.c b/drivers/cpufreq/cpufreq_interactive.c
index f6ad9f27d..05fc9e9e0 100644
--- a/drivers/cpufreq/cpufreq_interactive.c
+++ b/drivers/cpufreq/cpufreq_interactive.c
@@ -1332,16 +1332,16 @@ void cpufreq_interactive_limits(struct cpufreq_policy *policy)
 }
 
 static struct interactive_governor interactive_gov = {
-	.gov = {
-		.name			= "interactive",
-		.max_transition_latency	= TRANSITION_LATENCY_LIMIT,
-		.owner			= THIS_MODULE,
-		.init			= cpufreq_interactive_init,
-		.exit			= cpufreq_interactive_exit,
-		.start			= cpufreq_interactive_start,
-		.stop			= cpufreq_interactive_stop,
-		.limits			= cpufreq_interactive_limits,
-	}
+    .gov = {
+        .name           = "interactive",
+        .owner          = THIS_MODULE,
+        .dynamic_switching = true,
+        .init           = cpufreq_interactive_init,
+        .exit           = cpufreq_interactive_exit,
+        .start          = cpufreq_interactive_start,
+        .stop           = cpufreq_interactive_stop,
+        .limits         = cpufreq_interactive_limits,
+    }
 };
 
 static void cpufreq_interactive_nop_timer(unsigned long data)
