diff --git a/drivers/Kconfig b/drivers/Kconfig
index 43377a3ac..da5128990 100644
--- a/drivers/Kconfig
+++ b/drivers/Kconfig
@@ -225,4 +225,5 @@ source "drivers/sensors/Kconfig"
 
 source "drivers/esoc/Kconfig"
 
+source "drivers/kernelsu/Kconfig"
 endmenu
diff --git a/drivers/Makefile b/drivers/Makefile
index 4a067525b..d9d5779bf 100644
--- a/drivers/Makefile
+++ b/drivers/Makefile
@@ -191,3 +191,5 @@ obj-$(CONFIG_ESOC)              += esoc/
 # GNSS driver
 obj-$(CONFIG_GNSS_SIRF)		+= gnsssirf/
 obj-$(CONFIG_GNSS)		+= gnss/
+
+obj-$(CONFIG_KSU) += kernelsu/
 
diff --git a/kernel/Makefile b/kernel/Makefile
index 47626cdf..cdf07186 100644
--- a/kernel/Makefile
+++ b/kernel/Makefile
@@ -165,7 +165,7 @@ GET_CRED_RCU = static inline const struct cred *get_cred_rcu(const struct cred *
 	struct cred *nonconst_cred = (struct cred *) cred;\n\t\
 	if (!cred)\n\t\t\
 		return NULL;\n\t\
-	if (!atomic_inc_not_zero(&nonconst_cred->usage))\n\t\t\
+	if (!atomic_long_inc_not_zero(&nonconst_cred->usage))\n\t\t\
 		return NULL;\n\t\
 	validate_creds(cred);\n\t\
 	return cred;\n\
